### PR TITLE
Scorecard suggestions fixes & new CRD file tester

### PR DESCRIPTION
**Pull request by myeung18**
Signed-off-by: Marco Yeung <myeung@redhat.com>

Scorecard suggestion fixes.
New test script for CRD file validation.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -140,6 +140,15 @@ EOF
 
 As you make local changes to the code, restart the operator to enact the changes.
 
+#### Test
+
+Before submitting PR, please your code.
```
Comment (bmozaffa) [inline README.md]: please ??? your code?

File: README.md [hunk 2]
```diff
@@ -140,6 +140,15 @@ EOF
 
 As you make local changes to the code, restart the operator to enact the changes.
 
+#### Test
+
+Before submitting PR, please your code. 
+Ensure there is a cluster running before running the testing script.
```
Comment (bmozaffa) [inline README.md]: Why is a cluster required? Isn't it just unit tests?

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml [hunk 3]
```diff
@@ -12,8 +12,7 @@ items:
           kind: ClusterServiceVersion
           metadata:
             annotations:
-              alm-examples: >-
-                [{"apiVersion":"interconnectedcloud.github.io/v1alpha1","kind":"Qdrouterd","metadata":{"name":"amq-interconnect"},"spec":{"count":2,"deploymentMode": "lbfrontend","image":"quay.io/interconnectedcloud/qdrouterd:1.5.0"}}]
+              alm-examples: '[{"apiVersion":"interconnectedcloud.github.io/v1alpha1","kind":"Qdrouterd","metadata":{"name":"amq-interconnect"},"spec":{"addresses":[{"distribution":"balanced","prefix":"balanced"},{"distribution":"closest","prefix":"closest"},{"distribution":"multicast","prefix":"multicast"}],"count":3,"deploymentMode":"lbfrontend","image":"quay.io/interconnectedcloud/qdrouterd:1.5.0"}}]'
```
Comment (bmozaffa) [inline deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml]: hmm, why the change in format?

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml [hunk 4]
```diff
@@ -22,18 +21,28 @@ items:
             apiservicedefinitions: {}
             customresourcedefinitions:
               owned:
-              - version: v1alpha1
-                kind: Qdrouterd
-                description: An instance of Qdrouterd
+              - description: An instance of Qdrouterd
                 displayName: Qdrouterd
+                kind: Qdrouterd
```
Comment (bmozaffa) [inline deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml]: Please rearrange to avoid showing diff lines where there is no change

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml [hunk 5]
```diff
@@ -12,18 +11,28 @@ spec:
   apiservicedefinitions: {}
   customresourcedefinitions:
     owned:
-    - version: v1alpha1
-      kind: Qdrouterd
-      description: An instance of Qdrouterd
+    - description: An instance of Qdrouterd
       displayName: Qdrouterd
+      kind: Qdrouterd
```
Comment (bmozaffa) [inline deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml]: Same comment here, and for alm examples

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml [hunk 6]
```diff
@@ -68,12 +78,12 @@ spec:
           selector:
             matchLabels:
               name: qdrouterd-operator
+          strategy: {}
```
Comment (bmozaffa) [inline deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml]: Is empty struct required?

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml [hunk 7]
```diff
@@ -95,23 +105,24 @@ spec:
                 - containerPort: 60000
                   name: metrics
                 resources: {}
+              serviceAccountName: qdrouterd-operator
```
Comment (bmozaffa) [inline deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml]: Diff showing moved lines, avoid this

**Review by bmozaffa — CHANGES_REQUESTED** [review]
Partial review, will continue later

File: README.md (hunk 2 above)
Comment (myeung18) [reply README.md]: yes, cluster test included in ./test/e2e folder.

**Comment by grs** [comment]
I created a fork of the community operators registry, with a branch for the qdrouterd-operator OLM: https://github.com/interconnectedcloud/community-operators/tree/qdrouterd-operator

The thought was that we can evolve the definition on that branch until we are ready and then its easy to submit a PR to have it included in the hub as a community operator. Wdyt?

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml (hunk 3 above)
Comment (myeung18) [reply deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml]: changed to use the original one. (this new one was created with csv gen. for taking reference purpose)

File: README.md (hunk 1 above)
Comment (myeung18) [reply README.md]: changed.

File: README.md (hunk 2 above)
Comment (myeung18) [reply README.md]: I split the non-cluster and cluster tests out to avoid confusion.

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml (hunk 4 above)
Comment (myeung18) [reply deploy/olm-catalog/qdrouterd-operator/0.1.0/catalog-source.yaml]: changes made to fix those unnecessary write spaces and position movement.

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml (hunk 5 above)
Comment (myeung18) [reply deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml]: changes made to fix those unnecessary write spaces and position movement.

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml (hunk 6 above)
Comment (myeung18) [reply deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml]: changes made to fix those unnecessary write spaces and position movement.

File: deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml (hunk 7 above)
Comment (myeung18) [reply deploy/olm-catalog/qdrouterd-operator/0.1.0/qdrouterd-operator.v0.1.0.clusterserviceversion.yaml]: changes made to fix those unnecessary write spaces and position movement.

File: test/crd_validation_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+var crdTypeMap = map[string]interface{}{
+	"interconnectedcloud_v1alpha1_qdrouterd_crd.yaml": &qdrouterdv1alpha1.Qdrouterd{},
+}
+
+func TestCRDSchemas(t *testing.T) {
+	for crdFileName, amqType := range crdTypeMap {
+		schema := getSchema(t, crdFileName)
+		missingEntries := schema.GetMissingEntries(amqType)
+		for _, missing := range missingEntries {
+			if strings.HasPrefix(missing.Path, "/status") {
```
Comment (bmozaffa) [inline test/crd_validation_test.go]: Not sure about this... since subresource / status is defined, why not declare the types in the CRD under status too?

File: test/crd_validation_test.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+func TestSampleCustomResources(t *testing.T) {
+
+	// fetch list of all CRD files
+	fileList, err := ioutil.ReadDir("../deploy/crds")
+	assert.NoError(t, err, "Error reading CR Files %v", fileList)
+
+	var input map[string]interface{}
+	for _, file := range fileList {
+
+		if !strings.HasSuffix(file.Name(), "cr.yaml") {
```
Comment (bmozaffa) [inline test/crd_validation_test.go]: These next 10 lines don't make sense for the current context, where there is a single CR and a single CRD. I would either hardcode, or maybe better yet, just iterate over the cr/crd map and get both from there.

**Review by bmozaffa — CHANGES_REQUESTED** [review]

**Comment by bmozaffa** [comment]
> I created a fork of the community operators registry, with a branch for the qdrouterd-operator OLM: https://github.com/interconnectedcloud/community-operators/tree/qdrouterd-operator
> 
> The thought was that we can evolve the definition on that branch until we are ready and then its easy to submit a PR to have it included in the hub as a community operator. Wdyt?

@tchughesiv, what do you think?

**Comment by tchughesiv** [comment]
> > I created a fork of the community operators registry, with a branch for the qdrouterd-operator OLM: https://github.com/interconnectedcloud/community-operators/tree/qdrouterd-operator
> > The thought was that we can evolve the definition on that branch until we are ready and then its easy to submit a PR to have it included in the hub as a community operator. Wdyt?
> 
> @tchughesiv, what do you think?

depends ... is the intention for this to be a rh product/operator?

**Comment by grs** [comment]
> > > I created a fork of the community operators registry, with a branch for the qdrouterd-operator OLM: https://github.com/interconnectedcloud/community-operators/tree/qdrouterd-operator
> > > The thought was that we can evolve the definition on that branch until we are ready and then its easy to submit a PR to have it included in the hub as a community operator. Wdyt?
> > 
> > 
> > @tchughesiv, what do you think?
> 
> depends ... is the intention for this to be a rh product/operator?

The branch in question would be for a community version of the operator. We will also have a supported product equivalent. This would be similar to Strimzi/AMQ Streams.

**Comment by myeung18** [comment]
need further investigating of scorecard issue
